admittance_controller/src/admittance_controller.cpp

```diff
@@ -0,0 +1,505 @@
+      {
+        state_current.positions[joint_ind] = state_interfaces_[ind].get_value();
+        if (std::isnan(state_current.positions[joint_ind]))
+        {
+          state_current.positions = last_commanded_.positions;
+          break;
+        }
+      }
+      else if (inter_ind == state_vel_ind)
+      {
+        state_current.velocities[joint_ind] = state_interfaces_[ind].get_value();
+        if (std::isnan(state_current.positions[joint_ind]))
+        {
+          state_current.velocities = last_commanded_.velocities;
+          break;
+        }
+      }
+      else if (inter_ind == state_acc_ind)
+      {
+        state_current.accelerations[joint_ind] = state_interfaces_[ind].get_value();
+        if (std::isnan(state_current.positions[joint_ind]))
+        {
+          state_current.accelerations = last_commanded_.accelerations;
+          break;
+        }
+      }
+    }
+  }
+  force_torque_sensor_->get_values_as_message(ft_values);
+  if (
+    std::isnan(ft_values.force.x) || std::isnan(ft_values.force.y) ||
+    std::isnan(ft_values.force.z) || std::isnan(ft_values.torque.x) ||
+    std::isnan(ft_values.torque.y) || std::isnan(ft_values.torque.z))
+  {
+    ft_values = geometry_msgs::msg::Wrench();
+  }
+}
+
+void AdmittanceController::write_state_to_hardware(
+  const trajectory_msgs::msg::JointTrajectoryPoint & state_commanded)
+{
+  // Fill fields of state_reference argument from hardware state interfaces. If the hardware does
+  // not exist or the values are nan, the corresponding state field will be set to empty.
+
+  // if any interface has nan values, assume state_reference is the last command state
+  for (auto joint_ind = 0ul; joint_ind < num_joints_; joint_ind++)
+  {
+    for (auto inter_ind = 0ul; inter_ind < 3; inter_ind++)
+    {
+      auto ind = joint_ind + num_joints_ * inter_ind;
+      if (inter_ind == command_pos_ind)
+      {
+        command_interfaces_[ind].set_value(state_commanded.positions[joint_ind]);
+      }
+      else if (inter_ind == command_vel_ind)
+      {
+        command_interfaces_[ind].set_value(state_commanded.velocities[joint_ind]);
+      }
+      else if (inter_ind == command_acc_ind)
+      {
+        command_interfaces_[ind].set_value(state_commanded.accelerations[joint_ind]);
+      }
+    }
+  }
+  last_commanded_ = reference_admittance_;
+}
+
+void AdmittanceController::read_state_reference_interfaces(
+  trajectory_msgs::msg::JointTrajectoryPoint & state_reference)
+{
+  // TODO(destogl): check why is this here?
+  // Fill fields of state_reference argument from controller references.
+  // If the interface does not exist or
+  // the values are nan, the corresponding field will be set to empty
+
+  for (auto i = 0ul; i < position_reference_.size(); i++)
+  {
+    if (std::isnan(position_reference_[i]))
+    {
+      position_reference_[i].get() = last_reference_.positions[i];
+    }
+    state_reference.positions[i] = position_reference_[i];
```

Comment:
why not directly use storage from the message to write data from other controllers. It could be much simpler.

Reply:
I don't understand the comment, so I will try to explain this method and hopefully it will help. position_reference_ and velocity_reference_ point to the data that will be written to by other controllers via claiming the command interface. This method transfers that data to the state_reference which is used in the update loop. If nan values are written to the reference, then the last valid reference will be used.